packages/html-reporter/src/labelsView.tsx

```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import { navigate, Route } from './links';
+
+const testFilesRoutePredicate = (params: URLSearchParams) => !params.has('testId');
+const testCaseRoutePredicate = (params: URLSearchParams) => params.has('testId');
+
+export const LabelsView: React.FC<React.PropsWithChildren<{
+  labels: string[],
+}>> = ({ labels }) => {
+  if (!labels.length)
+    return null;
+
+  const encodedTags = labels.map(tag => encodeTag(tag));
+
+  return (
+    <>
+      {encodedTags.map(tag => (
+        <React.Fragment key={tag}>
+          <Route predicate={testFilesRoutePredicate}>
+            <LabelClick key={tag} tag={tag} />
+          </Route>
+          <Route predicate={testCaseRoutePredicate}>
+            <LabelLink key={tag} tag={tag} />
```

Comment:
These are seemingly the same, what is the difference and why do you use route here?

Reply:
There are two labels
 LabelLink - 'simple', inside <a/> and href - used for testCaseView
 LabelClick - 'smart', internal onClick handler that works in conjunction with filters by project or status (status AND label) and is used in testFileView

Reply:
I see. It looks like it is adding unnecessary complexity: labels control should not depend on the route. It would read simpler if you inlined smart and non-smart labels in the code that needs them respectively. Otherwise, you have 2 different clients that need 2 different controls, but you make them go through a single location that makes sense of the context via route and gives away the right control.